src/core/oracles/ChainlinkOracleDisputable.sol

```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {FixedPointMathLib} from "solmate/utils/FixedPointMathLib.sol";
+import {SafeCastLib} from "solmate/utils/SafeCastLib.sol";
+
+import {ChainlinkOracle} from "./ChainlinkOracle.sol";
+
+// constants and types
+import "../../config/errors.sol";
+
+/**
+ * @title ChainlinkOracleDisputable
+ * @author antoncoding
+ * @dev chainlink oracle that can be dispute by the owner
+ */
+contract ChainlinkOracleDisputable is ChainlinkOracle {
+    using SafeCastLib for uint256;
+
+    // base => quote => dispute period
+    mapping(address => mapping(address => uint256)) public disputePeriod;
+
+    /*///////////////////////////////////////////////////////////////
+                                 Events
+    //////////////////////////////////////////////////////////////*/
+
+    event DisputePeriodUpdated(address base, address quote, uint256 period);
+
+    /**
+     * @dev view function to check if dispute period is over
+     */
+    function isExpiryPriceFinalized(
+        address _base,
+        address _quote,
+        uint256 _expiry
+    ) external view returns (bool) {
+        return _isExpiryPriceFinalized(_base, _quote, _expiry);
+    }
+
+    /**
+     * @dev dispute the expiry price from the owner. Cannot dispute an un-reported price
+     * @param _base base asset
+     * @param _quote quote asset
+     * @param _expiry expiry timestamp
+     * @param _newPrice new price to set
+     */
+    function disputePrice(
+        address _base,
+        address _quote,
+        uint256 _expiry,
+        uint256 _newPrice
+    ) external onlyOwner {
+        ExpiryPrice memory entry = expiryPrices[_base][_quote][_expiry];
+        if (entry.reportAt == 0) revert OC_PriceNotReported();
+
+        if (entry.isDisputed) revert OC_PriceDisputed();
+
+        if (entry.reportAt + disputePeriod[_base][_quote] < block.timestamp) revert OC_DisputePeriodOver();
+
+        expiryPrices[_base][_quote][_expiry] = ExpiryPrice(true, uint64(block.timestamp), _newPrice.safeCastTo128());
+
+        emit ExpiryPriceUpdated(_base, _quote, _expiry, _newPrice, true);
+    }
+
+    /**
+     * @dev set the dispute period for a specific base / quote asset
+     * @param _base base asset
+     * @param _quote quote asset
+     * @param _period dispute period. Cannot be set to a vlue longer than 12 hours
+     */
+    function setDisputePeriod(
+        address _base,
+        address _quote,
+        uint256 _period
+    ) external onlyOwner {
+        if (_period > 12 hours) revert OC_InvalidDisputePeriod();
+
```

Comment:
@dsshap is 12 hours a reasonable number?

Reply:
I am trying to think if it's better to have flexibility or if it's not practical to give that sort of long dispute window... i lean towards not allowing more than a few hours, rather than half a day.

The options need to be exercised in a reasonable time relative to market swings. Locking up someone for (up to) 12 hours post expiry seem too much.

At the end of the day it's up to the creator of the oracle to decide what they want it to be. I wonder if we should have a protocol enforced maximum. What do you think about making `registerOracle` in grappa.sol enforce a protocol maximum before being accepted?